src/vm_cpu.rs

```diff
@@ -285,6 +285,79 @@ impl VmPmuState {
+        } else {
+            None
+        }
+    }
+
+    /// Saves the internal state for PMU counters. Stops started counters, and resets all configured counters.
+    /// This should be called in anticipation of an outbound context switch.
+    pub fn save_counters(&mut self) {
+        use crate::{println, CONSOLE_DRIVER};
+        use sbi::api::pmu;
+        use PmuCounterState::*;
+        for (i, state) in self.counter_state.iter_mut().enumerate() {
+            if matches!(state, Configured(_) | Started(_)) {
+                // Unwrap ok: Max value is 64
+                let counter_index = u64::try_from(i).unwrap();
+                if let Started(c) = state {
+                    if let Some(value) = Self::get_counter_csr_value(counter_index) {
+                        c.value = value;
+                    } else {
+                        println!("Warning: Failed to save value for {counter_index}");
+                    }
+                }
+                let result = pmu::stop_counters(
+                    counter_index,
+                    0x1,
+                    PmuCounterStopFlags::default().set_reset_flag(),
+                );
+                if result.is_err() && !matches!(result.err(), Some(sbi::Error::AlreadyStopped)) {
+                    println!("Warning: Failed to reset {counter_index}, {result:?}");
+                }
+            }
+        }
+    }
+
+    /// Restores PMU counters from the saved state, reconfiguring and starting them if necessary.
+    /// This should be called in anticipation of an outbound context switch.
+    pub fn restore_counters(&mut self) {
+        use crate::{println, CONSOLE_DRIVER};
+        use sbi::api::pmu;
+        use PmuCounterState::*;
+        for (i, state) in self.counter_state.iter().enumerate() {
+            match state {
+                Configured(c) | Started(c) => {
+                    // Unwrap ok: Max value is 64
+                    let counter_index = u64::try_from(i).unwrap();
+                    let result = pmu::configure_matching_counters(
+                        counter_index,
+                        0x1,
+                        c.config_flags.clear_skip_match().clear_clear_value(),
+                        c.event_type,
+                        c.event_data,
+                    );
+                    if result.is_ok() && matches!(state, Started(_)) {
+                        let start_flags = PmuCounterStartFlags::default().set_init_value();
+                        let result = pmu::start_counters(counter_index, 0x1, start_flags, c.value);
+                        if result.is_err() {
+                            println!("Warning: Failed to start {counter_index}, {result:?}");
+                        }
+                    } else if result.is_err() {
+                        println!("Warning: Failed to configure {counter_index}, {result:?}")
+                    }
```

Comment:
If we fail to start or configure a counter, should we put it in a "poisoned" state so that we don't end up exposing the CSR to the guest? e.g. all reads from a poisoned counter could trap and return 0.